app/assets/javascripts/discourse/app/components/add-category-tag-classes.js

```diff
@@ -1,57 +1,51 @@
 import Component from "@ember/component";
-import { observes } from "discourse-common/utils/decorators";
 import { scheduleOnce } from "@ember/runloop";
 
 export default Component.extend({
-  _slug: null,
+  tagName: "",
+  currentClasses: null,
 
-  didInsertElement() {
-    this._super(...arguments);
-    this.refreshClass();
+  init() {
+    this.currentClasses = new Set();
+    this._super();
   },
 
-  _updateClass() {
+  didReceiveAttrs() {
+    scheduleOnce("afterRender", this, this._updateClasses);
+  },
+
+  willDestroyElement() {
+    scheduleOnce("afterRender", this, this._removeClasses);
+  },
+
+  _updateClasses() {
     if (this.isDestroying || this.isDestroyed) {
       return;
     }
-    const slug = this.get("category.fullSlug");
-
-    this._removeClass();
 
-    const classes = [];
+    const desiredClasses = new Set();
 
+    const slug = this.category?.fullSlug;
     if (slug) {
-      classes.push("category");
-      classes.push(`category-${slug}`);
+      desiredClasses.add("category");
+      desiredClasses.add(`category-${slug}`);
     }
+    this.tags?.forEach((t) => desiredClasses.add(`tag-${t}`));
 
-    this.tags?.forEach((t) => classes.push(`tag-${t}`));
+    const addClasses = [...desiredClasses].filter(
+      (c) => !this.currentClasses.has(c)
+    );
+    const removeClasses = [...this.currentClasses].filter(
+      (c) => !desiredClasses.has(c)
+    );
 
-    document.body.classList.add(...classes);
-  },
+    document.body.classList.add(...addClasses);
```

Comment:
`classList.add()` does not duplicate classes, so you can use `...desiredClasses` here